### ISSUE TITLE
[QST] Inference logits mismatch vs training checkpoint — scaling_seqlen behavior differs between train/infer (intentional or bug?)

### ISSUE THREAD

**Issue by Lssyes** [NVIDIA/recsys-examples#260]
Hi NVIDIA recsys-examples maintainers,

I’m seeing a logits mismatch when running inference using a checkpoint trained with the provided training pipeline.

While investigating, I noticed a potential train/infer discrepancy when calling hstu_attn_2_cuda.varlen_fwd: the scaling_seqlen argument appears to be handled differently between inference and training.

In hstu_attn_interface.py, scaling_seqlen is overridden to max_seqlen_q when it is -1:
https://github.com/NVIDIA/recsys-examples/blob/ae93cfce90600ece15030902321e17a961595652/corelib/hstu/hstu_attn/hstu_attn_interface.py#L254

```
if scaling_seqlen == -1:
    scaling_seqlen = max_seqlen_q
```

However, during training, scaling_seqlen seems to remain the default value (-1) when calling the same varlen forward path (i.e., it is not explicitly set to max_seqlen_q). This could potentially make the attention scaling depend on different sequence length values between train and inference, leading to numerical differences in outputs.

Could you please clarify:
	1.	Is setting scaling_seqlen = max_seqlen_q in inference an intentional design choice (e.g., for stability/compatibility/performance)?
	2.	Or is it an unintended inconsistency/bug between training and inference code paths?

If it’s intended, I’d really appreciate any explanation of the rationale and the recommended way to keep train/infer behavior consistent (or how to configure scaling_seqlen properly).

**Comment by geoffreyQiu** [comment]
Hi,

1. This scaling factor does affect the result. 

> However, during training, scaling_seqlen seems to remain the default value (-1) when calling the same varlen forward path (i.e., it is not explicitly set to max_seqlen_q).

Training and inference should use the same code path for choosing the scaling factor. For each batch, it will be the largest seqlen_q by default. B


2. As for the mismatch between training and inference, did you run inference with kvcache? Currently we have difference casual masks (which is not aligned with the implementation from Meta, and is under fixing) esp. for contextual feature tokens, and this might cause the difference. Other useful data/info for me to look into this issue are also welcome.
